### ISSUE TITLE
Task panics hang

### ISSUE THREAD

**Issue by rsheeter** [googlefonts/fontc#90]
I tried to add a panic handler similar to the following:

```rust
                    let send = send.clone();
                    scope.spawn(move |_| {
                        let panicked = std::panic::catch_unwind(move || {
                            let result = work.exec(work_context);
                            if let Err(e) = send.send((id.clone(), result)) {
                                error!("Unable to write {:?} to completion channel: {}", id, e);
                            }
                        }).is_err();
                        
                        if panicked {
                            let result = Err(match &id {
                                AnyWorkId::Fe(..) => AnyWorkError::Fe(FeError::Panic),
                                AnyWorkId::Be(..) => AnyWorkError::Be(BeError::Panic),
                            });
                            if let Err(e) = send.send((id.clone(), result)) {
                                error!("Unable to write {:?} to completion channel: {}", id, e);
                            }
                        }
                    })
```

And got a 180 line error that would make a C++ compiler proud:

```
error[E0277]: the type `(dyn Work<fontir::orchestration::Context, WorkError> + Send + 'static)` may not be safely transferred across an unwind boundary
   --> fontc/src/main.rs:407:65
    |
407 |                           let panicked = std::panic::catch_unwind(move || {
    |  ________________________________________------------------------_^
    | |                                        |
    | |                                        required by a bound introduced by this call
408 | |                             let result = work.exec(work_context);
409 | |                             if let Err(e) = send.send((id.clone(), result)) {
410 | |                                 error!("Unable to write {:?} to completion channel: {}", id, e);
411 | |                             }
412 | |                         }).is_err();
    | |_________________________^ `(dyn Work<fontir::orchestration::Context, WorkError> + Send + 'static)` may not be safely transferred across an unwind boundary
    |
    = help: the trait `UnwindSafe` is not implemented for `(dyn Work<fontir::orchestration::Context, WorkError> + Send + 'static)`
    = note: required for `Unique<(dyn Work<fontir::orchestration::Context, WorkError> + Send + 'static)>` to implement `UnwindSafe`
    = note: required because it appears within the type `Box<(dyn Work<fontir::orchestration::Context, WorkError> + Send + 'static)>`
    = note: required because it appears within the type `AnyWork`
note: required because it's used within this closure
   --> fontc/src/main.rs:407:65
    |
407 |                         let panicked = std::panic::catch_unwind(move || {
    |                                                                 ^^^^^^^
note: required by a bound in `catch_unwind`
   --> ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:136:40
    |
136 | pub fn catch_unwind<F: FnOnce() -> R + UnwindSafe, R>(f: F) -> Result<R> {
    |                                        ^^^^^^^^^^ required by this bound in `catch_unwind`

error[E0277]: the type `(dyn Work<fontbe::orchestration::Context, fontbe::error::Error> + Send + 'static)` may not be safely transferred across an unwind boundary
   --> fontc/src/main.rs:407:65
    |
407 |                           let panicked = std::panic::catch_unwind(move || {
    |  ________________________________________------------------------_^
    | |                                        |
    | |                                        required by a bound introduced by this call
408 | |                             let result = work.exec(work_context);
409 | |                             if let Err(e) = send.send((id.clone(), result)) {
410 | |                                 error!("Unable to write {:?} to completion channel: {}", id, e);
411 | |                             }
412 | |                         }).is_err();
    | |_________________________^ `(dyn Work<fontbe::orchestration::Context, fontbe::error::Error> + Send + 'static)` may not be safely transferred across an unwind boundary
    |
    = help: the trait `UnwindSafe` is not implemented for `(dyn Work<fontbe::orchestration::Context, fontbe::error::Error> + Send + 'static)`
    = note: required for `Unique<(dyn Work<fontbe::orchestration::Context, fontbe::error::Error> + Send + 'static)>` to implement `UnwindSafe`
    = note: required because it appears within the type `Box<(dyn Work<fontbe::orchestration::Context, fontbe::error::Error> + Send + 'static)>`
    = note: required because it appears within the type `AnyWork`
note: required because it's used within this closure
   --> fontc/src/main.rs:407:65
    |
407 |                         let panicked = std::panic::catch_unwind(move || {
    |                                                                 ^^^^^^^
note: required by a bound in `catch_unwind`
   --> ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:136:40
    |
136 | pub fn catch_unwind<F: FnOnce() -> R + UnwindSafe, R>(f: F) -> Result<R> {
    |                                        ^^^^^^^^^^ required by this bound in `catch_unwind`

error[E0277]: the type `UnsafeCell<std::option::Option<Arc<fontir::ir::StaticMetadata>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   --> fontc/src/main.rs:407:65
    |
407 |                           let panicked = std::panic::catch_unwind(move || {
    |  ________________________________________------------------------_^
    | |                                        |
    | |                                        required by a bound introduced by this call
408 | |                             let result = work.exec(work_context);
409 | |                             if let Err(e) = send.send((id.clone(), result)) {
410 | |                                 error!("Unable to write {:?} to completion channel: {}", id, e);
411 | |                             }
412 | |                         }).is_err();
    | |_________________________^ `UnsafeCell<std::option::Option<Arc<fontir::ir::StaticMetadata>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
    |
    = help: within `lock_api::rwlock::RwLock<parking_lot::raw_rwlock::RawRwLock, std::option::Option<Arc<fontir::ir::StaticMetadata>>>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<std::option::Option<Arc<fontir::ir::StaticMetadata>>>`
    = note: required because it appears within the type `lock_api::rwlock::RwLock<parking_lot::raw_rwlock::RawRwLock, std::option::Option<Arc<fontir::ir::StaticMetadata>>>`
    = note: required for `Arc<lock_api::rwlock::RwLock<parking_lot::raw_rwlock::RawRwLock, std::option::Option<Arc<fontir::ir::StaticMetadata>>>>` to implement `UnwindSafe`
    = note: required because it appears within the type `fontir::orchestration::Context`
    = note: required because it appears within the type `AnyContext`
note: required because it's used within this closure
   --> fontc/src/main.rs:407:65
    |
407 |                         let panicked = std::panic::catch_unwind(move || {
    |                                                                 ^^^^^^^
note: required by a bound in `catch_unwind`
   --> ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:136:40
    |
136 | pub fn catch_unwind<F: FnOnce() -> R + UnwindSafe, R>(f: F) -> Result<R> {
    |                                        ^^^^^^^^^^ required by this bound in `catch_unwind`

error[E0277]: the type `UnsafeCell<HashMap<GlyphName, Arc<fontir::ir::Glyph>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   --> fontc/src/main.rs:407:65
    |
407 |                           let panicked = std::panic::catch_unwind(move || {
    |  ________________________________________------------------------_^
    | |                                        |
    | |                                        required by a bound introduced by this call
408 | |                             let result = work.exec(work_context);
409 | |                             if let Err(e) = send.send((id.clone(), result)) {
410 | |                                 error!("Unable to write {:?} to completion channel: {}", id, e);
411 | |                             }
412 | |                         }).is_err();
    | |_________________________^ `UnsafeCell<HashMap<GlyphName, Arc<fontir::ir::Glyph>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
    |
    = help: within `lock_api::rwlock::RwLock<parking_lot::raw_rwlock::RawRwLock, HashMap<GlyphName, Arc<fontir::ir::Glyph>>>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<HashMap<GlyphName, Arc<fontir::ir::Glyph>>>`
    = note: required because it appears within the type `lock_api::rwlock::RwLock<parking_lot::raw_rwlock::RawRwLock, HashMap<GlyphName, Arc<fontir::ir::Glyph>>>`
    = note: required for `Arc<lock_api::rwlock::RwLock<parking_lot::raw_rwlock::RawRwLock, HashMap<GlyphName, Arc<fontir::ir::Glyph>>>>` to implement `UnwindSafe`
    = note: required because it appears within the type `fontir::orchestration::Context`
    = note: required because it appears within the type `AnyContext`
note: required because it's used within this closure
   --> fontc/src/main.rs:407:65
    |
407 |                         let panicked = std::panic::catch_unwind(move || {
    |                                                                 ^^^^^^^
note: required by a bound in `catch_unwind`
   --> ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:136:40
    |
136 | pub fn catch_unwind<F: FnOnce() -> R + UnwindSafe, R>(f: F) -> Result<R> {
    |                                        ^^^^^^^^^^ required by this bound in `catch_unwind`

error[E0277]: the type `UnsafeCell<std::option::Option<Arc<fontir::ir::Features>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   --> fontc/src/main.rs:407:65
    |
407 |                           let panicked = std::panic::catch_unwind(move || {
    |  ________________________________________------------------------_^
    | |                                        |
    | |                                        required by a bound introduced by this call
408 | |                             let result = work.exec(work_context);
409 | |                             if let Err(e) = send.send((id.clone(), result)) {
410 | |                                 error!("Unable to write {:?} to completion channel: {}", id, e);
411 | |                             }
412 | |                         }).is_err();
    | |_________________________^ `UnsafeCell<std::option::Option<Arc<fontir::ir::Features>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
    |
    = help: within `lock_api::rwlock::RwLock<parking_lot::raw_rwlock::RawRwLock, std::option::Option<Arc<fontir::ir::Features>>>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<std::option::Option<Arc<fontir::ir::Features>>>`
    = note: required because it appears within the type `lock_api::rwlock::RwLock<parking_lot::raw_rwlock::RawRwLock, std::option::Option<Arc<fontir::ir::Features>>>`
    = note: required for `Arc<lock_api::rwlock::RwLock<parking_lot::raw_rwlock::RawRwLock, std::option::Option<Arc<fontir::ir::Features>>>>` to implement `UnwindSafe`
    = note: required because it appears within the type `fontir::orchestration::Context`
    = note: required because it appears within the type `AnyContext`
note: required because it's used within this closure
   --> fontc/src/main.rs:407:65
    |
407 |                         let panicked = std::panic::catch_unwind(move || {
    |                                                                 ^^^^^^^
note: required by a bound in `catch_unwind`
   --> ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:136:40
    |
136 | pub fn catch_unwind<F: FnOnce() -> R + UnwindSafe, R>(f: F) -> Result<R> {
    |                                        ^^^^^^^^^^ required by this bound in `catch_unwind`

error[E0277]: the type `UnsafeCell<std::option::Option<Arc<Vec<u8>>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   --> fontc/src/main.rs:407:65
    |
407 |                           let panicked = std::panic::catch_unwind(move || {
    |  ________________________________________------------------------_^
    | |                                        |
    | |                                        required by a bound introduced by this call
408 | |                             let result = work.exec(work_context);
409 | |                             if let Err(e) = send.send((id.clone(), result)) {
410 | |                                 error!("Unable to write {:?} to completion channel: {}", id, e);
411 | |                             }
412 | |                         }).is_err();
    | |_________________________^ `UnsafeCell<std::option::Option<Arc<Vec<u8>>>>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
    |
    = help: within `lock_api::rwlock::RwLock<parking_lot::raw_rwlock::RawRwLock, std::option::Option<Arc<Vec<u8>>>>`, the trait `RefUnwindSafe` is not implemented for `UnsafeCell<std::option::Option<Arc<Vec<u8>>>>`
    = note: required because it appears within the type `lock_api::rwlock::RwLock<parking_lot::raw_rwlock::RawRwLock, std::option::Option<Arc<Vec<u8>>>>`
    = note: required for `Arc<lock_api::rwlock::RwLock<parking_lot::raw_rwlock::RawRwLock, std::option::Option<Arc<Vec<u8>>>>>` to implement `UnwindSafe`
    = note: required because it appears within the type `fontbe::orchestration::Context`
    = note: required because it appears within the type `AnyContext`
note: required because it's used within this closure
   --> fontc/src/main.rs:407:65
    |
407 |                         let panicked = std::panic::catch_unwind(move || {
    |                                                                 ^^^^^^^
note: required by a bound in `catch_unwind`
   --> ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:136:40
    |
136 | pub fn catch_unwind<F: FnOnce() -> R + UnwindSafe, R>(f: F) -> Result<R> {
    |                                        ^^^^^^^^^^ required by this bound in `catch_unwind`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `fontc` due to 6 previous errors
```

**Comment by rsheeter** [comment]
`todo!()` in async work is unfun because of this

**Comment by rsheeter** [comment]
Fixed by #371
